### PR TITLE
Improve dind documentation and fix dind testing

### PR DESCRIPTION
**Pull request by marun**
The contribution doc may not be the best place for this, so placement suggestions are welcome.

### CODE REVIEW

File: CONTRIBUTING.adoc [hunk 1]
```diff
@@ -193,6 +193,103 @@ Follow these steps to ensure that virtual box interfaces are unmanaged:
...
+        $ vagrant up
+        $ vagrant ssh
+        $ cd /vagrant
+        $ hack/dind-cluster.sh start
+
+These steps will create a new VM, update its packages, ensure that
+docker is installed, and launch a new OpenShift dind cluster.  Package
+installation and the building of the dind image will take some time,
+but are a one-time cost.  Subsequent cluster launches should be
+considerably faster.
```
Comment (danwinship) [inline CONTRIBUTING.adoc]: So does "dind-cluster.sh start" pull in the latest openshift binary and sdn scripts? Or do you mean "it's a one-time cost if you plan to do a bunch of testing against a particular checkout of origin"?

File: CONTRIBUTING.adoc [hunk 2]
```diff
@@ -193,6 +193,103 @@ Follow these steps to ensure that virtual box interfaces are unmanaged:
...
+
+Whether using dind-cluster.sh or networking.sh to run tests, it's
+possible to override the default test regexes via the
+NETWORKING_E2E_FOCUS and NETWORKING_E2E_SKIP environment variables.
+These variables set the '-focus' and '-skip' arguments supplied to the
+https://github.com/onsi/ginkgo[ginkgo] test runner.
+
+==== Running Kubernetes e2e tests
+
+It's possible to target the Kuberntes e2e tests against a running
```
Comment (danwinship) [inline CONTRIBUTING.adoc]: typo "Kuberntes"

File: CONTRIBUTING.adoc (hunk 1 above)
Comment (marun) [inline CONTRIBUTING.adoc]: "dind-cluster.sh start" creates containers that mount the repo, builds the binaries and configuration in the master container as part of the master deploy, and then deploys to the node containers.  It's very similar to how the vagrant deploy works (equivalent to 'vagrant up').  The 'one-time cost' I'm referring to is the package installation on the vm (dnf update && dnf install docker...) and the base dind image (dnf update && dnf install openvswitch docker...).

Would it be useful to make that explicit?  e.g. "Package installation on the VM and building the dind image" and "openshift will be built as part of cluster creation"?

File: CONTRIBUTING.adoc (hunk 2 above)
Comment (marun) [inline CONTRIBUTING.adoc]: done.

File: CONTRIBUTING.adoc (hunk 1 above)
Comment (danwinship) [inline CONTRIBUTING.adoc]: Yeah, maybe split it up into two parts, with everything up to "vagrant up" being explained as one-time setup, and then ssh+cd+start as the iteratable part.

File: CONTRIBUTING.adoc [hunk 3]
```diff
@@ -193,6 +193,103 @@ Follow these steps to ensure that virtual box interfaces are unmanaged:
...
+use cases.
+
+To run a dind cluster in a VM, follow steps 1-3 of the Vagrant
+instructions and then execute the following:
+
+        $ export OPENSHIFT_DIND_DEV_CLUSTER=true
+        $ vagrant up
+        $ vagrant ssh
+        $ cd /vagrant
+        $ hack/dind-cluster.sh start
```
Comment (danwinship) [inline CONTRIBUTING.adoc]: ...
    Error response from daemon: Conflict. The name "openshift-master" is already in use by container 564ed56cf240. You have to delete (or rename) that container to be able to reuse that name.
    json: cannot unmarshal array into Go value of type types.ContainerJSON
    Error response from daemon: Conflict. The name "openshift-node-1" is already in use by container f5d7f21ce0de. You have to delete (or rename) that container to be able to reuse that name.
    json: cannot unmarshal array into Go value of type types.ContainerJSON

does it automatically get started after the first "vagrant up"? Or are those errors spurious? Or...?

**Comment by marun** [comment]
@danwinship Hopefully the latest pathset addresses your concerns.  I've also enhanced the vagrant provisioning to allow dind cluster interaction directly from the VM (it had previously been necessary to login to one of the containers).

**Comment by danwinship** [comment]
lgtm

**Comment by marun** [comment]
This latest patchset cleans up vagrant deployment and ensures that the networking tests can be executed.

If you've already deployed a vagrant vm, you can either recreate the vm or run the following steps to run the tests:

    # Login to vagrant vm
    $ vagrant ssh

    # Set the gopath
    $ export GOPATH=/vagrant/_output/local/go

    # Add the gopath to the user's profile
    $ echo ${GOPATH} >> ~/.bash_profile

    # Install ginkgo
    go get github.com/onsi/ginkgo/ginkgo

    # Restart the cluster
    hack/dind-cluster restart

    # Run the tests
    hack/dind-cluster test-net-e2e

**Comment by marun** [comment]
[test][extended:networking]

**Comment by marun** [comment]
[test]

File: CONTRIBUTING.adoc [hunk 4]
```diff
@@ -196,6 +196,117 @@ Follow these steps to ensure that virtual box interfaces are unmanaged:
...
+openshift.  Assuming the 'vagrant up' command completes without error,
+a dind OpenShift cluster should now be running on the VM.  To access
+the cluster, login to the VM:
+
+        $ vagrant ssh
+
+Once on the VM, the 'oc' and 'openshift' commands can be used to
+interact with the cluster:
+
+        $ oc get nodes
```
Comment (danwinship) [inline CONTRIBUTING.adoc]: FWIW, most oc commands seem to work, but "oc project foo" gives an error. ("oc new-project foo" creates the project successfully, but then gives an error when trying to switch to it):

    [vagrant@openshift-dind-host vagrant]$ oc new-project foo
    error: open /tmp/openshift-dind-cluster/openshift/openshift.local.config/master/admin.kubeconfig: permission denied

**Comment by danwinship** [comment]
otherwise lgtm

**Comment by marun** [comment]
@danwinship Latest patchset should fix the permission issue.

**Comment by marun** [comment]
[test]

**Comment by marun** [comment]
This PR allows the extended networking tests to run.  Getting them passing will be a subsequent effort.

**Comment by dcbw** [comment]
LGTM

**Comment by eparis** [comment]
Am I understanding correctly that the extended network testing is running, but failing now? Yeah, we can't merge something we know will break CI for everyone else. Maybe I'm misunderstanding the openshift-jenkins/test failure:
```
Directory /var/lib/docker/vfs/dir does not exist, skipping.
Error while running ssh/sudo command: 
set -e
pushd /data/src/github.com/openshift//origin >/dev/null
export PATH=$GOPATH/bin:$PATH

echo '***************************************************'
echo 'Running test/extended/networking.sh...'
time test/extended/networking.sh
echo 'Finished test/extended/networking.sh'
echo '***************************************************'

popd >/dev/null
        
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!
```

**Comment by eparis** [comment]
[test]

**Comment by eparis** [comment]
[test]

**Comment by eparis** [comment]
[merge]

**Pull request merged**